File: 백준/Silver/24446. 알고리즘 수업 － 너비 우선 탐색 3/알고리즘 수업 － 너비 우선 탐색 3.cc
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

int N, M, R;
queue<int>q;

void bfs(vector<vector<int>>& graph, vector<int>& visited,int start) {
	visited[start] = 0;
	q.push(start);

	while (!q.empty()) {
		int now = q.front();
		q.pop();

		for (int i = 0;i < graph[now].size();i++) {
			int next = graph[now][i];
			if (visited[next] == -1) {
				visited[next] = visited[now] + 1;
				q.push(next);
			}
		}
	}
}



int main() {
	cin >> N >> M >> R;

	vector<vector<int>> graph(N+1);
	vector<int> visited(N + 1, -1);


	for (int i = 0;i < M;i++) {
		int s, e;
		cin >> s >> e;
		graph[s].push_back(e);
		graph[e].push_back(s);
	}

	bfs(graph,visited,R);
	for (int i = 1;i <= N;i++) {
		cout << visited[i] << '\n';
	}

	return 0;
}

File: 백준/Silver/24446. 알고리즘 수업 － 너비 우선 탐색 3/알고리즘 수업 － 너비 우선 탐색 3.py
import sys
from collections import deque

input = sys.stdin.readline


def bfs(start):
    q = deque()
    q.append(start)
    visited[start] = 0
    while q:
        now = q.popleft()
        for i in dic.get(now, {}):
            if visited[i] == -1:
                visited[i] = visited[now] + 1
                q.append(i)


N, M, R = map(int, input().split())
visited = [-1] * (N + 1)
dic = {}
for i in range(M):
    s, e = map(int, input().split())
    if s not in dic:
        dic[s] = [e]
    else:
        dic[s].append(e)
    if e not in dic:
        dic[e] = [s]
    else:
        dic[e].append(s)

bfs(R)

for i in range(1, N + 1):
    print(visited[i])

